SAM4E8E Port

FHeilmann opened the pull request:
Here is my second attempt of a SAM4e8e port based only on CMSIS which has a friendlier license. The SAM4e8e can, for example, be found on the popular DuetWifi and DuetEthernet 3D printer boards.

The port is based on the sam3x CMSIS branch so that would have to be merged first to get the CMSIS core files.

The following things are working (although I have not stress-tested them yet):

Timers, watchdog, etc.
- Serial (over UART0)
- GPIO 
- ADC via the Analog Frontend Controller (AFEC)
- SPI via both USARTs in SPI mode (this was of interest since the TMC2660 on the duet wifi use one of the USART-SPIs)

ToDo:
- Serial over USB (I'm currently looking into the LPC176x branch to see if I can get it to work)
- True hardware SPI

@KevinOConnor I'd like to perform the stepping benchmarks, can you share your method? If I should modify anything about the PR please let me know!

Bests,

Florian

john-- commented:
Beautiful

KevinOConnor commented:
Thanks!  I'll take a look over the next couple of days.

> Serial over USB (I'm currently looking into the LPC176x branch to see if I can get it to work)

FYI, I recently added support for USB on the SAMD21.  The LPC176x usb controller is a bit goofy - you may find the SAMD21 code is a better reference.

> I'd like to perform the stepping benchmarks, can you share your method?

Take a look at issue #271 - I measur via console.py tests.

FHeilmann commented:
@KevinOConnor thanks for the hints, I'll look into benchmarking later.

Regarding the USB, it looks as if the SAMD0 USB is just as different from the SAM4 USB device stuff as the LPC176x USB. I'll have to spend some time on it

KevinOConnor commented:
Thanks!  The commits look good to me.  The only thing I noticed is that lib/README needs to be updated.

> Regarding the USB, it looks as if the SAMD0 USB is just as different from the SAM4 USB device stuff as the LPC176x USB.

Indeed. And SAM3 USB is totally different.  The USB on the Atmel AT90 line is also unique.  The STM32F103 has yet another implementation..

That said, though, from glancing at the spec, the SAM4 USB implementation looked sane.

-Kevin

KevinOConnor commented:
Thanks.  One note - when I go to http://packs.download.atmel.com/ I see a `Atmel.SAM4E_DFP.1.1.57.atpack` file, but I don't see a `Atmel.SAM4E_DFP.1.1.57 (1).atpack` file.  Is the site doing something funky, or is that a typo.  Also, was the file pulled on June 8th or August 6th?

Impressive benchmark!  Just for my own reference, what did you get for the `ticks` parameter for the one stepper, two steppers, and three stepper cases?

Finally, I'm good with committing.  Or did you want me to wait while you work on additional stuff?

-Kevin

FHeilmann commented:
That was a typo ofc. I fixed it and squashed some minor commits so the PR consists of meaningful commits.

I ran the procedure as you suggested in the other PR (pasted the first section once and then pasted the remainder over and over with different tick values until I get shutdowns).

The values I obtained were the following:

3 steppers: 330
2 steppers: 281
1 stepper: 329

I tested each one of them 3-4 times in sequence to make sure there were no errors.

FHeilmann commented:
One more thing thing. The makefile references my newer bossac build that people using the the main repo have to obtain first. Do you have a suggestion how to incorporate the downloading and compiling of the the latest version bossa version into klipper? Maybe during the install script? I can provide step by step terminal commands. I'm sure the sam3 and sam21 port will be able to use this newer version as well.

FHeilmann commented:
Thanks for merging! Let me know if you want me to provide the respective sections on flashing and benchmarking of the SAM4e as well.

KevinOConnor commented:
> Let me know if you want me to provide the respective sections on flashing and benchmarking of the SAM4e as well.

Yes, it would be great if you could provide the benchmark details (in particular the gcc version and the init sequence).  If you have info on the SAM4 bootloader / flashing process, that would be great too.

-Kevin

FHeilmann commented:
I'm closing this for now, I'll submit a separate PR for the documentation changes.
